Readme.txt
GBMailer
Supported platforms: Win95/98/2000/NT4.0, Linux-i386
September 2000 version 2.02

Author:
Goran Boban
gboban70@gmail.com
Home page: https://github.com/gboban/GBMailer

License: GPL (please read LICENSE.TXT)


	1. Description
	--------------

GBMailer is command-line mailer wich will send an text file as mail.
It is written to be executed from CGI-scripts, but it can be used for 
any other purpose.

See history.txt for changes in this version.

NOTE:	Be sure to read Bugs section (section 2.) of this readme file
	before mailing bug reports or questions about using GBMailer

USAGE:

gbmail [-v] -to <address list...> [-i] [-file <filename>] [-h <smtp.server.name>]
	[-p <port_number>] [-from <address>] [-cc <address list...>]
	[-bcc <address list>] [-s <Subject>] [-t <timeout>]


OPTIONS:

	-v	verbose output.

	-file	Name of text file to send as message body.
		(This is not required since version 1.12)

	-i	Get input from stdin (i.e: redirection and pipes).

	-h	Name of SMTP server to connect. If this option is not set
		GBMailer will assume 'localhost'.

	-p	Port number for SMTP protocol. Port is set to 25 by default.

	-from	An optional e-mail ddress of sender. If not set GBMailer
		will try to assume an e-mail address by combinating name of
		current user and name of local machine.
		You should set this option in most cases.

	-to	REQUIRED	E-mail addres (or addresses) of recipient(s).
				At least one address must cast after -to switch.

	-cc	(carbon copy) E-mail addresses of recipients which will be listed
		in Cc: field.

	-bcc	(blind carbon copy) E-mail addresses of recipients which you don't
		want to be listed in message which will be sent.

	-s	Subject of message.

	-t	Timeout for socket operations (in seconds). Set to 300 secs by
		default.

RETURN VALUE:
	There are three groups of errors eturned by GBMailer, which calling
	application can examine to learn details about failure:

	- error code <0 is negative value of error returned by sockets
	subsystem (WSAGetLastError() on Win or errno/h_errno on Linux).
	Note: error codes returned on Windows systems and linux systems
	will be different. Windows socket errors start at 10000 (WSABASEER)
	while Linux error codes are in range 1-123.

	- error code = 1 inicates missing command-line switch, missing
	switch parameter or missing file given after -file switch.

	- error code in range 200-600 is value of SMTP response (see RFC821).
	Note that even 2NN and 3NN may be returned, even if response codes
	in this range are generally not errors. This is because GBMailer
	will signal error on every unexpected rsponse to SMTP commands.

	Finally, if return value is 0, this means no error (i.e. mail sent).


EXAMPLE:

gbmail -file message.txt -h smtp.server.com -from me@my.machine.com -to you@your.machine.com -s "Hello there"

or if you want to use redirection

gbmail -i -h smtp.server.com -from me@my.machine.com -to you@your.machine.com -s "Hello there" < message.txt


	2. Bugs
	-------

	Please send bug reports and sugestions to:

	gboban70@gmail.com

	IMPORTANT NOTE: When sending bug reports or questions on using GBMailer please
	be sure to include following information:

		- Program name (GBMailer in this case)
		- Program version (2.02)
		- OS you are using
		- Version of OS you are using (you can include build number and service
		pack number)
		- How did you invoke gbmail.exe (from command-line or from script, if
		from script please send me name of interpreter and version)
		- Problem you want to report

	You may include any other information you find relevant (or you think it may be
	relevant) for this problem.
	Also be sure to use correct e-mail address (in From: or Reply to: field) so I
	can send you answer.




HISTORY.TXT
History file for GBMailer version 2.02

version 2.02 (september 2000)

	- Fixed problem with ESMTP multiline responses to HELO command
	(Solution proposed by Christian Marg (einhirn@gmx.de))

source published under GPL license (September 2000)

	- This is same old 2.01 version, but now distributed with source

version 2.01 (Jun 2000)

	- Bug fix.

version 2.00 (Jun 2000)

	- Linux version

	- Now you can use redirection and pipes (see -i switch).
	
	- socket errors are now returned as negative values (<0),
	SMTP responses as positive (2NN-5NN) and if return value
	is 1 than this is switch error. No error is still 0.

Version 1.12 (August 1999)

	- Now, you can send e-mail without body (i.e. -file switch is no more
	required).
	
	- Internal structure is cleared and some typos in error messages are
	corrected.

Version 1.11 (May 1998)

	- Corrected SMTP sintax.

Version 1.10 (February 1998)

	- Added -bcc switch, so now you can use it to pass "blind carbon copy"
	list.

	- "To:" field is now created by GBMailer, so now recipient will  always
	get "To:" field in header.

Version 1.01 (January 1998)

	- Problem with -s option is now corrected, so you should be able
	to send mail with subject.

	- There is no 255 characters limit on argument to -file option any more.



smtp.c
/* 
*	Copyright (C) 2000  Goran Boban
*
*	last revision date: 13 Sep 2000 23:59:50
*
*
*    This program is free software; you can redistribute it and/or modify
*    it under the terms of the GNU General Public License as published by
*    the Free Software Foundation; either version 2 of the License, or
*    (at your option) any later version.
*
*    This program is distributed in the hope that it will be useful,
*    but WITHOUT ANY WARRANTY; without even the implied warranty of
*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*    GNU General Public License for more details.
*
*    You should have received a copy of the GNU General Public License
*    along with this program; if not, write to the Free Software
*    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*
*/
#include "smtp.h"
#include "gbsock.h"

#include <stdlib.h>

/* ======================================================= CONSTANTS */

static char XMailer[]="GBMailer version 2.02"; /* mailer identifier */


/* =============================================== sendMail(...) helpers */

/* smtp_errorCode(...)
*		Converts firts tree characters of string to integer.
*		(SMTP error codes are returned on the begining of line)
*/
static int smtp_errorCode(char* str){

	return atoi(str);
}


/* sendFile(...)
*		Reads file block by block and writes contents to socket
*
*		- returns 0 if no error
*		- returns <0 on socket error
*/
static int sendFile(SOCKET sock, FILE* fp, unsigned long timeout){
	char fbuf[SMTP_BUFLEN];
	int bytesRead;
	int err;

	do{
		bytesRead = fread(fbuf, 1, SMTP_BUFLEN - 1, fp);
		fbuf[bytesRead] = 0;
		err=sock_write(sock, fbuf, timeout);
		if(err < 0) return err;
	}while(bytesRead == SMTP_BUFLEN - 1);

	return 0;
}

/* smtp_sendCommand(...)
*		sends SMTP command
*
*		if return value is >0 -> SMTP response code
*		if return value is <0 -> socket error
*
*	REMARKS: caller should check if returned SMTP code
*	is error or not.
*/
static int smtp_sendCommand(SOCKET sock, char* cmd, unsigned long timeout){
	char buf[SMTP_BUFLEN];
	int err;

	/* send command */
	err=sock_write(sock, cmd, timeout);
	if(err < 0) return err; /* socket error? */

	/* get response */
	err=sock_readLine(sock, buf, SMTP_BUFLEN, timeout);
	if(err < 0) return err; /* socket error? */

	/* return SMTP error */
	return smtp_errorCode(buf);
}

/* smtp_nameRecipients(...)
*		Iterates through list of recipients (To, Cc or Bcc recipients) and
*		sends "RCPT TO:" SMTP command for each
*
*		- returns 0 if no error
*		- returns >0 on SMTP error
*		- returns <0 on socket error
*/
static int smtp_nameRecipients(SOCKET sock, char** rcptList, unsigned long timeout){
	char buf[SMTP_BUFLEN];
	int err;
	int i;

	for(i = 0; rcptList[i]; ++i){	/* for each */
		/* send command */
		sprintf(buf, "RCPT TO: <%s>\r\n", rcptList[i]);
		err = smtp_sendCommand(sock, buf, timeout);
		if((err != 250) && (err != 251)) return err;
	}

	return 0; /* OK */
}

/*	smtp_sendCtrlHeaders(...)
*		sends set of control headers, FROM:, RCPT TO: ...
*
*		- returns 0 if no error
*		- returns >0 on SMTP error
*		- returns <0 on socket error
*/
static int smtp_sendCtrlHeaders(SOCKET sock, PMAIL_MSG msg, unsigned int timeout){
	char buf[SMTP_BUFLEN];	
	int err;

	/* MAIL FROM: */
	sprintf(buf, "MAIL FROM: <%s>\r\n", msg->from);
	err = smtp_sendCommand(sock, buf, timeout);
	if(err != 250) return err;

	/* RCPT TO: (TO) */
	err = smtp_nameRecipients(sock, msg->to, timeout);
	if(err) return err;

	if(msg->cc){	/* RCPT TO: (CC) */
		err = smtp_nameRecipients(sock, msg->cc, timeout);
		if(err) return err;
	}

	if(msg->bcc){	/* RCPT TO: (BCC) */
		err = smtp_nameRecipients(sock, msg->bcc, timeout);
		if(err) return err;
	}

	return 0;
}

/* smtp_sendData(...)
*		sends message contents
*
*		- returns 0 if no error
*		- returns >0 on SMTP error
*		- returns <0 on socket error
*/
static int smtp_sendData(SOCKET sock, PMAIL_MSG msg, unsigned long timeout){
	char buf[SMTP_BUFLEN];
	int oldVerbose;
	int err;
	int i;

	/* start data transfer */
	err = smtp_sendCommand(sock, "DATA\r\n", timeout);
	if(err != 354) return err;

	/* disable verbose mode */
	oldVerbose = sock_verbose;
	sock_verbose = 0;


	/* identify mailer */
	sprintf(buf, "X-Mailer: %s\r\n", XMailer);
	err=sock_write(sock, buf, timeout);
	if(err < 0) return err;

	/* To: */
	err = sock_write(sock, "To: ", timeout);
	if(err < 0) return err;
	for(i = 0; msg->to[i]; ++i){
		if(i == 0) sprintf(buf, "%s", msg->to[i]);
		else sprintf(buf, ", %s", msg->to[i]);
		err = sock_write(sock, buf, timeout);
		if(err < 0) return err;
	}
	sprintf(buf, "\r\n");
	err=sock_write(sock, buf, timeout);
	if(err < 0) return err;


	/* Cc: */
	if(msg->cc){
		err = sock_write(sock, "Cc: ", timeout);
		if(err < 0) return err;
		for(i = 0; msg->cc[i]; ++i){
			if(i == 0) sprintf(buf, "%s", msg->cc[i]);
			else sprintf(buf, ", %s", msg->cc[i]);
			err = sock_write(sock, buf, timeout);
			if(err < 0) return err;
		}
		sprintf(buf, "\r\n");
		err=sock_write(sock, buf, timeout);
		if(err < 0) return err;
	}

	/* From: */
	sprintf(buf, "From: %s\r\n", msg->from);
	err=sock_write(sock, buf, timeout);
	if(err < 0) return err;

	/* Subject: */
	if(msg->subject){
		sprintf(buf, "Subject: %s\r\n", msg->subject);
		err=sock_write(sock, buf, timeout);
		if(err < 0) return err;
	}

	sprintf(buf, "\r\n");
	err=sock_write(sock, buf, timeout);
	if(err < 0) return err;

	/* send message body */
	if(msg->src == isFILE){
		err = sendFile(sock, msg->fp, timeout);
		if(err) return err;
	}

	/* restore verbose mode */
	sock_verbose = oldVerbose;

	/* "." */
	err = smtp_sendCommand(sock, "\r\n.\r\n", timeout);
	if(err != 250) return err;

	return 0;
}


/* smtp_connect(...)
*		connets to SMTP host
*
*		- returns 0 if no error
*		- returns >0 on SMTP error
*		- returns <0 on socket error
*/
static int smtp_connect(SOCKET sock, unsigned long addr, unsigned short port, unsigned long timeout){
	char buf[SMTP_BUFLEN];
	int smtpError;
	int err, i;

	/* connect socket */
	err = sock_connect(sock, addr, port);
	if(err < 0) return err;

	/* get greeting message */
	err=sock_readLine(sock, buf, SMTP_BUFLEN, timeout);
	if(err < 0) return err;

	/* check SMTP error code (is it SMTP server at all?) */
	smtpError = smtp_errorCode(buf);
	if(smtpError != 220)	return smtpError;

	/* introduce ourself (HELO) */
	sprintf(buf, "HELO %s\r\n", sock_getHostName());
	err = smtp_sendCommand(sock, buf, timeout);
	if(err != 250) return err;

	/* Solution proposed by Christian Marg (einhirn@gmx.de)
	 *	see sock_dataReady(SOCKET sock) in gbsock.c
	 */
   /* if it is a ESMTP-Host, read out the ehlo-responses */
   while ((i=sock_dataReady(sock))!=0) {
      //printf("Still having %d bytes to read\n",i);
      err=sock_readLine(sock,buf,SMTP_BUFLEN,timeout);
      if(err < 0) return err; /* socket error? */
      err=smtp_errorCode(buf);
      if (err!=250) return err;
  }

	return 0; /* OK */
}

/* smtp_sendMessage(...)
*
*
*		- returns 0 if no error
*		- returns >0 on SMTP error
*		- returns <0 on socket error
*/
static int smtp_sendMessage(SOCKET sock, PMAIL_MSG msg, unsigned long timeout){
	int err;

	/* send control headers (i.e: MAIL FROM:, RCPT TO: ...) */
	err = smtp_sendCtrlHeaders(sock, msg, timeout);

	/* if no error, then send data (actual message) */
	if(!err)	err = smtp_sendData(sock, msg, timeout);

	return err;
}

/* ====================================================== sendMail(...) */

/*	sendMail(...)
*		Sends message using given SMTP server.
*
*		- returns 0 if no error
*		- returns >0 on SMTP error
*		- returns <0 on socket error
*/
int sendMail(PMAIL_MSG msg, PSMTP_HOST host, unsigned long timeout){
	SOCKET sock;
	unsigned long hostAddr;
	int err;

	/* translate host name to address */
	hostAddr = sock_getHostByName(host->name, &err);
	if(!hostAddr) return err;

	/* create socket */
	sock = sock_createSocket(&err);
	if(sock == INVALID_SOCKET) return err;

	/* connect to SMTP server */
	err = smtp_connect(sock, hostAddr, host->port, timeout);

	if(!err){
		/* we are connected - seond message */
		err = smtp_sendMessage(sock, msg, timeout);
		if(!err){
			err = smtp_sendCommand(sock, "QUIT\r\n", timeout);
			if(err == 221) err = 0; /* not error */
		}
	}

	sock_close(sock);

	return err;
}



smtp.h
/* 
*	Copyright (C) 2000  Goran Boban
*
*	last revision date: 04 Jun 2000 00:24:44
*
*
*    This program is free software; you can redistribute it and/or modify
*    it under the terms of the GNU General Public License as published by
*    the Free Software Foundation; either version 2 of the License, or
*    (at your option) any later version.
*
*    This program is distributed in the hope that it will be useful,
*    but WITHOUT ANY WARRANTY; without even the implied warranty of
*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*    GNU General Public License for more details.
*
*    You should have received a copy of the GNU General Public License
*    along with this program; if not, write to the Free Software
*    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*
*/
#ifndef _GBSMTP_H_
#define _GBSMTP_H_

#include <stdio.h>

/* define default size for bufers used in SMTP conversation
* 512 to comply with rfc821 + 1 byte, so we can null-terminate buffer
*/
#define SMTP_BUFLEN 513

typedef enum{						/* where are we going to look for input? */
	isNOInput,					/* message won't have a body */
	isFILE,						/* message body should be read from file */
} INPUT_SOURCE;

/* Description of message */
typedef struct{
	char* from;			/* senders address ("From:" header of e-mail message) */
	char* subject;		/* subject of message */
	char** to;			/* list of recipients ("To:") */
	char** cc;			/*	list of recipients ("Cc:") */
	char** bcc;			/*	list of recipients ("Bcc:") */
	FILE* fp;			/* Handle of file from which to read message body */
	INPUT_SOURCE src;	/* where are we going to look for input? */
} MAIL_MSG, *PMAIL_MSG;

/* Description of SMTP server */
typedef struct{
	char* name;					/* server name */
	unsigned short port;		/* server port */
} SMTP_HOST, *PSMTP_HOST;

/*	sendMail(...)
*		Sends message using given SMTP server.
*
*		- returns 0 if no error
*		- returns >0 on SMTP error
*		- returns <0 on socket error
*/
int sendMail(PMAIL_MSG msg, PSMTP_HOST host, unsigned long timeout);

#endif


os_socket.h
/* 
*	Copyright (C) 2000  Goran Boban
*
*	last revision date: 07 Sep 2000 04:03:29 PM
*
*
*    This program is free software; you can redistribute it and/or modify
*    it under the terms of the GNU General Public License as published by
*    the Free Software Foundation; either version 2 of the License, or
*    (at your option) any later version.
*
*    This program is distributed in the hope that it will be useful,
*    but WITHOUT ANY WARRANTY; without even the implied warranty of
*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*    GNU General Public License for more details.
*
*    You should have received a copy of the GNU General Public License
*    along with this program; if not, write to the Free Software
*    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*
*/
#ifndef _OS_SOCKS_H_
#define _OS_SOCKS_H_

#ifdef _TGT_WIN32_
	/* we compile for WIN32 platform */
#include <winsock.h>

#define ETIMEDOUT WSAETIMEDOUT

#else
	/* linux expected */

#include <errno.h>
#include <sys/time.h>
#include <unistd.h>
/*#include <sys/socket.h>*/
#include <netinet/in.h>
#include <netdb.h>

#include <sys/ioctl.h>

#define closesocket close
#define WSAGetLastError() errno
#define ioctlsocket ioctl

#define SOCKET_ERROR 	-1
#define INVALID_SOCKET	-1

#define SOCKET long int

typedef struct sockaddr *LPSOCKADDR;

#endif

#endif


main.c
/* 
*	Copyright (C) 2000  Goran Boban
*
*	last revision date: 04 Jun 2000 00:24:21
*
*
*    This program is free software; you can redistribute it and/or modify
*    it under the terms of the GNU General Public License as published by
*    the Free Software Foundation; either version 2 of the License, or
*    (at your option) any later version.
*
*    This program is distributed in the hope that it will be useful,
*    but WITHOUT ANY WARRANTY; without even the implied warranty of
*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*    GNU General Public License for more details.
*
*    You should have received a copy of the GNU General Public License
*    along with this program; if not, write to the Free Software
*    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*
*/
#include "gbsock.h"
#include "smtp.h"

#include <stdio.h>
#include <stdlib.h>

/* ======================================================= CONSTANTS */

/* switch names */
char switchVerbose[]	= "-v";		/* verbose output */
char switchFile[] 	= "-file";	/* -file [input_file] */
char switchStdin[]	= "-i";		/* read input from stdin */
char switchHost[]		= "-h";		/* SMTP host name */
char switchPort[]		= "-p";		/* SMTP port number */
char switchTo[]		= "-to";		/* list of "To:" recipients */
char switchCc[]		= "-cc";		/* list of "Cc:" recipients */
char switchBcc[]		= "-bcc";	/* list of "Bcc:" recipients */
char switchFrom[]		= "-from";	/* -from [e-mail_address] */
char switchSubject[]	= "-s";		/* -s "subject of the message" */
char switchTimeout[]	= "-t";		/* timeout for socket operations */

char usage[]="\ngbmail \t[-v] -to <addres> [-file <filename>] [-i] [-h <smtp_server_name>]\n\t[-p <port_number>] [-from <addres>] [-cc <Cc:>] [-bcc <Bcc:>] \n\t[-s <Subject:>] [-t <timeout>]\n";

char argumentRequired[] = "GBMailer: switch requires argument: %s\n";
char missingSwitch[]		= "GBMailer: missing switch: %s\n";
char noMemory[]			= "GBMailer: no enough memory\n";

/* ========================================================= GLOBALS */

char from[SMTP_BUFLEN]="";			/* for storing senders address */
unsigned long timeout=300;		/* to hold timeout value (-t switch) */
char *fname = 0;					/* file name if -file switch is given */

MAIL_MSG msg = {			/* message description */
	0,				/* char* -> no sender (From:) */
	0,				/* char* -> no subject (Subject:) */
	0,				/* char** -> no recipients (To:) */
	0,				/* char** -> no Cc: */
	0,				/* char** -> no Bcc: */
	0,				/* FILE* no input file */
	isNOInput	/* where are we going to look for input? */
};

SMTP_HOST host = {	/* SMTP host description */
	"localhost", 		/* char* -> host name - assume localhost */
	25 					/* unsigned short -> SMTP port */
};


/* ======================================================= FUNCTIONS */

#ifdef _TGT_WIN32_
/*	cuserid(...)
*		Fills buf with login name of current user.
*
*		- if succesfull, returns buf
*		- returns 0 on failure
*
*		REMARKS:
*		For windows, we define this function, on other platforms
*		(Linux/UNIX) it should already exist.
*/
char* cuserid(char* buf){
	unsigned long buflen = 256;

	if(GetUserName(buf, &buflen)){
		return buf;
	}else return 0;
}
#endif

/*	testarg(...)
*		Itereates through command-line arguments to check for presence
*		of unknown switches
*
*		- if no unknown switches (i.e. no error) than returns 0
*		- if unknown switch is found, than returns 1
*/
int testargs(int argc, char* argv[]){
	int i;

	for(i=0;i<argc;++i){
		if(argv[i][0]=='-'){
			if(!strcmp(argv[i], switchVerbose)) continue;
			if(!strcmp(argv[i], switchFile)) continue;
			if(!strcmp(argv[i], switchStdin)) continue;
			if(!strcmp(argv[i], switchHost)) continue;
			if(!strcmp(argv[i], switchPort)) continue;
			if(!strcmp(argv[i], switchTo)) continue;
			if(!strcmp(argv[i], switchCc)) continue;
			if(!strcmp(argv[i], switchBcc)) continue;
			if(!strcmp(argv[i], switchFrom)) continue;
			if(!strcmp(argv[i], switchSubject)) continue;
			if(!strcmp(argv[i], switchTimeout)) continue;
			printf("GBMailer: unrecognized switch - %s\n", argv[i]);
			return 1;
		}
	}

	return 0;
}

/* findswitch(...)
*		Finds position of command-line switch s in argv[]
*
*		- if switch is found, then index of switch is returned
*		- if switch is not found, returns 0 (we don't expect switches at
*		argv[0])
*/
int findswitch(char *s, int argc, char* argv[]){
	int i;

	/* iterate through argument-list until we find
	*	switch or run out of arguments
	*/
	for(i=1;(i<argc)&&(strcmp(s, argv[i]));++i);

	if(i >= argc) return 0; /* switch not found */
	
	return i;	/* switch is at argv[i] */
}

/*	countSwitchArguments(...)
*		
*		- if switch is found, then count non-switch arguments after it
*		(i.e. all arguments before next switch or end of argv[] list)
*		- if no switch - return 0
*/
int countSwitchArguments(char* s, int argc, char* argv[]){
	int first;
	int last;

	first = findswitch(s, argc, argv);
	if(!first) return 0;

	for(last = first + 1; (last < argc) && (argv[last][0]!='-');++last);
	
	return (last - first - 1);
}

/* copyUpToNextSwitch(...)
*		Copies all elements of src[] starting from start to dest[]
*		(starting at 0), until it founds string beginning with "-"
*		(i.e. switch) or reaches end of src[]
*
*		- No return value
*/
void copyUpToNextSwitch(char* dest[], char* src[], int start, int length){
	int i;

	for(i = 0; i < length; ++i){
		dest[i] = src[start + i];
	}
}

/* processVerboseSwitch(...)
*		Checks if -v switch is present in argv[]
*
*		- if -v switch is found, than sets verbose
*		variable in msgs.c to nonzero
*		- if -v switch is not found - does nothing
*
*		Always returns 0 (i.e. function never fails)
*/
int processVerboseSwitch(int argc, char* argv[]){

	if(findswitch(switchVerbose, argc, argv)){
		sock_verbose=1;
	}

	return 0;
}

/* processStdinSwitch(...)
*		Checks if -i (read message from stdin) switch is
*		present in argv[]
*
*		- if -i switch is found, than sets inputSource
*		GLOBAL variable to isSTDIN
*		- if -i switch is not found - does nothing
*
*		Always returns 0 (i.e. function never fails)
*/
int processStdinSwitch(int argc, char* argv[]){

	if(findswitch(switchStdin, argc, argv)){
		msg.src = isFILE;
		msg.fp = stdin;
	}

	return 0;
}

/* processToSwitch(...)
*		Checks for presence of -to switch in command-line params
*		and initializes (char**) msg.to if one found
*
*		- if switch found, returns 0 (no error) and initializes
*		to member of msg GLOBAL variable (message descriptor)
*		- if no -to switch or no arguments to switch, returns
*		non-zero (error)
*/
int processToSwitch(int argc, char* argv[]){
	int i, count;

	/* switch exists? */
	if(!(i=findswitch(switchTo, argc, argv))){
		printf(missingSwitch, switchTo);
		return 1;
	}

	if(((i+1) >= argc) || (argv[i+1][0]=='-')){
		printf(argumentRequired, switchTo);
		return 1;
	}

	/* OK we have switch at i, now allocate and fill toList */
	count = countSwitchArguments(switchTo, argc, argv);

	msg.to = (char**)malloc((count+1) * sizeof(char*));
	if(!msg.to){
		printf(noMemory);
		return 1;
	}
	
	msg.to[count] = 0;
	copyUpToNextSwitch(msg.to, argv, i + 1, count);

	return 0;
}

/* processCcSwitch(...)
*		Checks for presence of -cc switch in command-line params
*		and initializes (char**) msg.cc if one found
*
*		- if switch found, returns 0 (no error) and initializes
*		cc member of msg GLOBAL variable (message descriptor)
*		- if no switch returns 0 - this is not an error
*		- if no arguments to switch, returns non-zero (error)
*/
int processCcSwitch(int argc, char* argv[]){
	int i, count;

	if((i=findswitch(switchCc, argc, argv))){

		if(((i+1) >= argc) || (argv[i+1][0]=='-')){
			printf(argumentRequired, switchCc);
			return 1;
		}

		/* OK we have switch at i, now allocate and fill ccList */
		count = countSwitchArguments(switchCc, argc, argv);

		msg.cc = (char**)malloc((count+1) * sizeof(char*));
		if(!msg.cc){
			printf(noMemory);
			return 1;
		}
		
		msg.cc[count] = 0;
		copyUpToNextSwitch(msg.cc, argv, i + 1, count);
	}

	return 0;
}

/* processBccSwitch(...)
*		Checks for presence of -bcc switch in command-line params
*		and initializes (char**) msg.bcc if one found
*
*		- if switch found, returns 0 (no error) and initializes
*		bcc member of msg GLOBAL variable (message descriptor)
*		- if no switch returns 0 - this is not an error
*		- if no arguments to switch, returns non-zero (error)
*/
int processBccSwitch(int argc, char* argv[]){
	int i, count;

	if((i=findswitch(switchBcc, argc, argv))){

		if(((i+1) >= argc) || (argv[i+1][0]=='-')){
			printf(argumentRequired, switchBcc);
			return 1;
		}

		/* OK we have switch at i, now allocate and fill bccList */
		count = countSwitchArguments(switchBcc, argc, argv);

		msg.bcc = (char**)malloc((count+1) * sizeof(char*));
		if(!msg.bcc){
			printf(noMemory);
			return 1;
		}
	
		msg.bcc[count] = 0;
		copyUpToNextSwitch(msg.bcc, argv, i + 1, count);
	}

	return 0;
}

/* processTimeoutSwitch(...)
*		Checks if -t switch is present in argv[]
*		and sets timeout, if yes
*
*		- if -t switch is found, than sets timeout
*		GLOBAL variable to value given on command-line
*		- if switch is found, but without proper argument,
*		returns non-zero (error)
*		- if -t switch is not found - does nothing
*		(this is not an error)
*/
int processTimeoutSwitch(int argc, char* argv[]){
	int i;

	if((i=findswitch(switchTimeout, argc, argv))){

		if(((i+1) >= argc) || (argv[i+1][0]=='-')){
			printf(argumentRequired, switchTimeout);
			return 1;
		}

		timeout=atoi(argv[i+1]);
	}
	return 0;
}

/* processFileSwitch(...)
*		Checks if -file switch is present in argv[]
*		and sets fname and inputSource, if yes
*
*		- if -file switch is found, than sets fname
*		GLOBAL variable to value given on command-line
*		- if switch is found, but without proper argument,
*		returns non-zero (error)
*		- if -file switch is not found - does nothing
*		(this is not an error)
*/
int processFileSwitch(int argc, char* argv[]){
	int i;

	if((i=findswitch(switchFile, argc, argv))){

		if(((i+1) >= argc) || (argv[i+1][0]=='-')){
			printf(argumentRequired, switchFile);
			return 1;
		}

		fname=argv[i+1];
		msg.src = isFILE;

		msg.fp=fopen(fname, "rb");
		if(!msg.fp){
			printf("GBMailer: failed to open file: %s\n", fname);
			return 1;
		}

	}
	return 0;
}

/* processFromSwitch(...)
*		Checks if -from switch is present in argv[]
*		and sets msg.from GLOBAL variable
*
*		- if -from switch is found, than sets from member
*		of msg GLOBAL variable to value given on command-line
*		- if switch is found, but without proper argument,
*		returns non-zero (error)
*		- if -from switch is not found - constructs from in form:
*		user_name@localhost_name
*/
int processFromSwitch(int argc, char* argv[]){
	char buf[512];
	int i;

	if((i=findswitch(switchFrom, argc, argv))){

		if(((i+1) >= argc) || (argv[i+1][0]=='-')){
			printf(argumentRequired, switchFrom);
			return 1;
		}

		msg.from = argv[i+1];
	}else{
		sprintf(from, "%s@%s", cuserid(buf), sock_getHostName());
		msg.from = from;
	}
	return 0;
}

/* processHostSwitch(...)
*		Checks if -h switch is present in argv[]
*		and sets host.name GLOBAL variable
*
*		- if -h switch is found, than sets name member
*		of host GLOBAL variable to value given on command-line
*		- if switch is found, but without proper argument,
*		returns non-zero (error)
*		- if -h switch is not found - does nothing
*		(this is not an error)
*/
int processHostSwitch(int argc, char* argv[]){
	int i;

	if((i=findswitch(switchHost, argc, argv))){

		if(((i+1) >= argc) || (argv[i+1][0]=='-')){
			printf(argumentRequired, switchHost);
			return 1;
		}

		host.name = argv[i+1];
	}
	return 0;
}

/* processPortSwitch(...)
*		Checks if -p switch is present in argv[]
*		and sets host.port GLOBAL variable
*
*		- if -p switch is found, than sets port member
*		of host GLOBAL variable to value given on command-line
*		- if switch is found, but without proper argument,
*		returns non-zero (error)
*		- if -p switch is not found - does nothing
*		(this is not an error)
*/
int processPortSwitch(int argc, char* argv[]){
	int i;

	if((i=findswitch(switchPort, argc, argv))){

		if(((i+1) >= argc) || (argv[i+1][0]=='-')){
			printf(argumentRequired, switchPort);
			return 1;
		}

		host.port=atoi(argv[i+1]);
	}
	return 0;
}

/* processSubjectSwitch(...)
*		Checks if -s switch is present in argv[]
*		and sets msg.subject GLOBAL variable
*
*		- if -s switch is found, than sets subject member
*		of msg GLOBAL variable to value given on command-line
*		- if switch is found, but without proper argument,
*		returns non-zero (error)
*		- if -s switch is not found - does nothing
*		(this is not an error)
*/
int processSubjectSwitch(int argc, char* argv[]){
	int i;

	if((i=findswitch(switchSubject, argc, argv))){

		if(((i+1) >= argc) || (argv[i+1][0]=='-')){
			printf(argumentRequired, switchSubject);
			return 1;
		}

		msg.subject = argv[i+1];
	}
	return 0;
}

/* procargs(...)
*		Calls specific processing functions for
*		each switch
*
*		- if any of processXXXSwitch function fails
*		procarg returns 1
*		- if no error, returns 0
*/
int procargs(int argc, char* argv[]){

	/* any unknown switches? */
	if(testargs(argc, argv)) return 1;

	/* VERBOSE (-v) */
	if(processVerboseSwitch(argc, argv)) return 1;

	/* READ FROM STDIN (-i) */
	if(processStdinSwitch(argc, argv)) return 1;

	/* TO (-to) */
	if(processToSwitch(argc, argv)) return 1;

	/* TIMEOUT (-t) */
	if(processTimeoutSwitch(argc, argv)) return 1;

	/* FILE (-file) */
	if(processFileSwitch(argc, argv)) return 1;

	/* FROM (-from) */
	if(processFromSwitch(argc, argv)) return 1;

	/* HOST (-h) */
	if(processHostSwitch(argc, argv)) return 1;

	/* PORT (-p) */
	if(processPortSwitch(argc, argv)) return 1;

	/* CC (-cc) */
	if(processCcSwitch(argc, argv))	return 1;

	/* BCC (-bcc) */
	if(processBccSwitch(argc, argv)) return 1;

	/* SUBJECT (-s) */
	if(processSubjectSwitch(argc, argv)) return 1;

	return 0;
}

void showError(int errorCode){
	
	if(errorCode == -ETIMEDOUT) 								/* timeout */
		printf("GBMailer: connection timed out\n");	
	else if(errorCode > 0)		/* SMTP error */
		printf("GBMailer: SMTP error: %d\n", errorCode);
	else															/* socket error */
		printf("GBMailer: socket error: %d\n", -errorCode);
}

/* ============================================================ MAIN */
int main(int argc, char *argv[]){
	int err;		/* to store error codes */

	/* initialize sockets (Windoze WSAStartup()) in gbsock.c */
	err=sock_startup();
	if(err < 0){	/* couldn't initialize sockets */
		printf("GBMailer: Failed to initialize sockets (%d)\n", err);
		return err;
	}

	/* pass through command-line arguments and initialize
	* coresponding variables, show usage message if anything
	* wrong with arguments
	*/
	err = procargs(argc, argv);
	if(!err){

		/* send message */
		err = sendMail(&msg, &host, timeout);
		if(!err){	/* We did it :) */
			printf("GBMailer: Mail sent!\n");
		}else{		/* Failed */
			showError(err);
			printf("GBMailer: failed to send mail\n");
		}
	}else{	/* error in command-line arguments */
		printf(usage);
	}

	/* release resources */
	if(msg.fp) fclose(msg.fp);
	if(msg.to) free(msg.to);
	if(msg.cc) free(msg.cc);
	if(msg.bcc) free(msg.bcc);

	/* sockets cleanup (Windoze WSACleanup()) in gbsock.c */
	sock_cleanup();

	return err;
}

